dali/pipeline/operator/sequence_shape.h

```diff
@@ -0,0 +1,298 @@
+  }
+
+  inline size_t NumChannels(size_t sample_idx) const {
+    assert(sample_idx < NumSamples());
+    assert(channels_dim_ < NumDimsToExpand());
+    return dims_to_expand_[sample_idx][channels_dim_];
+  }
+
+  inline const TensorListShape<> &DimsToExpand() const {
+    return dims_to_expand_;
+  }
+
+  inline TensorLayout Layout() const {
+    return layout_;
+  }
+
+  inline TensorLayout ExpandedLayout() const {
+    return layout_.first(num_expand_dims_);
+  }
+
+ private:
+  TensorLayout layout_;
+  int frames_dim_;
+  int channels_dim_;
+  bool expand_frames_;
+  bool expand_channels_;
+  int num_expand_dims_;
+  bool is_channel_first_;
+  TensorListShape<> dims_to_expand_;
+  size_t num_expanded_;
+};
+
+namespace sequence_utils {
+
+template <typename FrameRange>
+class SliceIterator {
+  using IndexType = typename FrameRange::IndexType;
+  using SliceViewType = typename FrameRange::SliceViewType;
+
+ public:
+  SliceIterator(const FrameRange &range, IndexType idx) : range_{range}, idx_{idx} {}
+
+  SliceViewType operator*() const {
+    return range_[idx_];
+  }
+
+  bool operator==(const SliceIterator &other) {
+    return idx_ == other.idx_;
+  }
+
+  bool operator!=(const SliceIterator &other) {
+    return !(*this == other);
+  }
+
+  void operator++() {
+    ++idx_;
+  }
+
+ private:
+  const FrameRange &range_;
+  IndexType idx_;
+};
+
+struct SliceView {
+  uint8_t *ptr;
+  TensorShape<> shape;
+  size_t type_size;
+};
+
+class UnfoldedSliceRange {
+ public:
+  using IndexType = size_t;
+  using SliceViewType = SliceView;
+
+  inline UnfoldedSliceRange(SliceViewType view, int ndims_to_unfold)
+      : view_{view},
+        num_slices_{[&]() {
+          assert(view.shape.sample_dim() >= ndims_to_unfold);
+          auto vol = volume(view.shape.begin(), view.shape.begin() + ndims_to_unfold);
+          return static_cast<std::make_unsigned_t<decltype(vol)>>(vol);
+        }()},
+        slice_shape_{view.shape.begin() + ndims_to_unfold, view.shape.end()},
+        slice_stride_{view.type_size * volume(slice_shape_)} {}
+
+  inline SliceIterator<UnfoldedSliceRange> begin() const {
+    return {*this, 0};
+  }
+
+  inline SliceIterator<UnfoldedSliceRange> end() const {
+    return {*this, NumSlices()};
+  }
+
+  inline SliceViewType operator[](IndexType idx) const {
+    return {view_.ptr + idx * SliceSize(), SliceShape(), view_.type_size};
+  }
+
+  inline size_t NumSlices() const {
+    return num_slices_;
+  }
+
+  inline TensorShape<> SliceShape() const {
+    return slice_shape_;
+  }
+
+  inline size_t SliceSize() const {
+    return slice_stride_;
+  }
+
+ private:
+  SliceViewType view_;
+  size_t num_slices_;
+  TensorShape<> slice_shape_;
+  size_t slice_stride_;
+};
+
+template <typename Backend>
+struct TensorVectorBuilder {
+  TensorVectorBuilder(int num_samples, DALIDataType type, int sample_dim, TensorLayout layout,
+                      bool is_pinned, AccessOrder order)
```

Comment:
I don't fully grasp, what's the purpose of this class? IIUC, it get's some `TensorVector` related configuration options and creates a `TensorVector` from the `SliceView`. In the same time, it's a one-off function, since it moves the `tv_`.
What would happen if user does something like:
```c++
TensorVectorBuilder tvb(...);
tvb.push(sview1);
auto tv = tvb.take();
tvb.push(sview2);
tvb.take();
tvb.take();
```
And the last question, that comes to my mind is, how it is different from a function:
```c++
TensorVector<Backend> create(int num_samples, DALIDataType type, int sample_dim, TensorLayout layout, bool is_pinned, AccessOrder order, const SliceView& view);
```

Reply:
It is a utility to build the tensorvector, you first create the instance of it, then push `num_samples` times the SliceViews and take the result, i.e. a build tensorvector once. This is something that happens in at least three different places of the code of SequenceOperator. Separating direct interaction with TV API greatly shortened those functions, assured in all cases I don't forget about setting some property of tensorvector and moved direct interaction with tensorvector api into a single place, which is quite useful given the fact that @klecki is reworking it.

Reply:
How about putting some `DALI_ENFORCE` there? I mean, it would make me really mad, if I would need to debug a segfault coming from double-take call.

Reply:
I don't like the idea of introducing runtime check on using some internal utility. It's not that by deriving from SequenceOperator you can override some method in such a way that some instance of this utility will be used incorrectly - its usage is limited to the scope of a call to a method. It feels like adding DALI_ENFORCE on loop invariant. How about adding docs with explicitly mentioning that you can call take once and only when there are exactly num_samples (which I added an assert for).